### PR TITLE
[SPARK-36733][SQL] Fix a perf issue in SchemaPruning when a struct has many fields

### PR DESCRIPTION
**Pull request by sarutak**
### What changes were proposed in this pull request?

This PR fixes a perf issue in `SchemaPruning` when a struct has many fields (e.g. >10K fields).
The root cause is `SchemaPruning.sortLeftFieldsByRight` does N * M order searching.
```
 val filteredRightFieldNames = rightStruct.fieldNames
    .filter(name => leftStruct.fieldNames.exists(resolver(_, name))) 
```

To fix this issue, this PR proposes to use `HashMap` to expect a constant order searching.
This PR also adds `case _ if left == right => left` to the method as a short-circuit code.

### Why are the changes needed?

To fix a perf issue.

### Does this PR introduce _any_ user-facing change?

No. The logic should be identical.

### How was this patch tested?

I confirmed that the following micro benchmark finishes within a few seconds.
```
import org.apache.spark.sql.catalyst.expressions.SchemaPruning
import org.apache.spark.sql.types._

var struct1 = new StructType()
(1 to 50000).foreach { i =>
  struct1 = struct1.add(new StructField(i + "", IntegerType))
}

var struct2 = new StructType()
(50001 to 100000).foreach { i =>
  struct2 = struct2.add(new StructField(i + "", IntegerType))
}

SchemaPruning.sortLeftFieldsByRight(struct1, struct2)
SchemaPruning.sortLeftFieldsByRight(struct2, struct2)
```

The correctness should be checked by existing tests.

### CODE REVIEW

**Review by HyukjinKwon — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SchemaPruning.scala [hunk 1]
```diff
@@ -65,16 +67,19 @@ object SchemaPruning extends SQLConfHelper {
...
-        val filteredRightFieldNames = rightStruct.fieldNames
-          .filter(name => leftStruct.fieldNames.exists(resolver(_, name)))
-        val sortedLeftFields = filteredRightFieldNames.map { fieldName =>
-          val resolvedLeftStruct = leftStruct.find(p => resolver(p.name, fieldName)).get
-          val leftFieldType = resolvedLeftStruct.dataType
-          val rightFieldType = rightStruct(fieldName).dataType
-          val sortedLeftFieldType = sortLeftFieldsByRight(leftFieldType, rightFieldType)
-          StructField(fieldName, sortedLeftFieldType, nullable = resolvedLeftStruct.nullable,
-            metadata = resolvedLeftStruct.metadata)
+        val leftStructTreeMap =
```
Comment (taroplus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SchemaPruning.scala:70]: based on my use-cases, often times left and right are the same thing, can we have a condition like
```
case (_, _) if left == right => left
```
i'm not sure if we need to recreate the whole thing for those situations

**Comment by dongjoon-hyun** [comment]
cc @sunchao

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 2]
```diff
@@ -3768,6 +3768,24 @@ class SQLConf extends Serializable with Logging {
...
+   */
+  private[sql] def comparator: Comparator = {
+    if (caseSensitiveAnalysis) {
+      org.apache.spark.sql.catalyst.analysis.caseSensitiveComparator
+    } else {
+      org.apache.spark.sql.catalyst.analysis.caseInsensitiveComparator
+    }
+  }
+
+  private[sql] val fieldNameOrdering = new Ordering[String] {
```
Comment (taroplus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:3783]: not sure how much difference it makes, however checking config (caseSensitiveAnalysis) for every evaluation doesn't look necessary, can we do create an ordering per config ...
```
  private[sql] def fieldNameOrdering: Ordering[String] = {
    if (caseSensitiveAnalysis) {
      (a: String, b: String) => a.compareTo(b)
    } else {
      (a: String, b: String) => a.compareToIgnoreCase(b)
    }
  }
```

Comment (sarutak) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:3783]: Thank you. It's reasonable.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SchemaPruning.scala [hunk 3]
```diff
@@ -65,16 +67,19 @@ object SchemaPruning extends SQLConfHelper {
...
+          TreeMap(leftStruct.map(_.name).zip(leftStruct): _*)(conf.fieldNameOrdering)
```
Comment (sunchao) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SchemaPruning.scala:71]: I wonder if we can use a `HashMap` for this:

```scala
  private def formatFieldName(name: String): String =
    if (conf.caseSensitiveAnalysis) name else name.toLowerCase(Locale.ROOT)

  ...

  val leftStructTreeMap =
    HashMap(leftStruct.map(f => formatFieldName(f.name)).zip(leftStruct): _*)
```

Comment (sarutak) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SchemaPruning.scala:71]: Seems better. Thanks.

**Comment by HyukjinKwon** [comment]
retest this please

**Review by sunchao — APPROVED** [review]
LGTM!

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**